pageserver/src/layered_repository.rs

```diff
@@ -0,0 +1,1056 @@
+        // Note: We release all the layer structs, and start fresh
+        // with an empty map. This keeps memory usage in check, but is
+        // probably too aggressive. Some kind of LRU policy would be
+        // appropriate.
+        //
+
+        // Call freeze() for any unfrozen layers (that is, layers that
+        // haven't been written to disk yet)
+        // Call unload() for all layers, to release memory.
+        //
+        // FIXME: We do this by creating a whole new LayerMap. I couldn't
+        // figure out the borrowing rules to remove and insert entries
+        // to the old LayerMap while iterating through it.
+        let old_layers = std::mem::take(&mut *layers);
+        for layer in old_layers.inner.values() {
+            if !layer.is_frozen() {
+                let new_layers = layer.freeze(last_valid_lsn, &*self.walredo_mgr)?;
+
+                for new_layer in new_layers {
+                    info!("freeze returned {} {}-{}", new_layer.get_relish_tag(), new_layer.get_start_lsn(), new_layer.get_end_lsn());
+                    layers.insert(Arc::clone(&new_layer));
+                }
+            } else {
+                // FIXME: if this fails, we're in trouble because we already
+                // swapped the 'layers' with the empty one. Hence panic on error.
+                layer.unload().expect("could not unload layer from memory");
+                layers.insert(Arc::clone(layer));
+            }
+        }
+
+        // Also save the metadata, with updated last_valid_lsn and last_record_lsn, to a
+        // file in the timeline dir
+        let ancestor_timelineid = if let Some(x) = &self.ancestor_timeline {
+            Some(x.timelineid)
+        } else {
+            None
+        };
+
+        let metadata = TimelineMetadata {
+            last_valid_lsn: self.last_valid_lsn.load(),
+            last_record_lsn: self.last_record_lsn.load(),
+            prev_record_lsn: self.prev_record_lsn.load(),
+            ancestor_timeline: ancestor_timelineid,
+            ancestor_lsn: self.ancestor_lsn,
+        };
+        LayeredRepository::save_metadata(self.conf, self.timelineid, self.tenantid, &metadata)?;
+
+        // If there were any concurrent updates on the timeline, we would have to work
+        // harder to make sure we don't lose the new updates. Currently, that shouldn't
+        // happen, because the WAL receiver process is responsible for both updating
+        // the timeline and calling checkpoint()
+        assert!(self.last_valid_lsn.load() == last_valid_lsn);
+
+        Ok(())
+    }
+
+    /// Remember that WAL has been received and added to the page cache up to the given LSN
+    fn advance_last_valid_lsn(&self, lsn: Lsn) {
+        let old = self.last_valid_lsn.advance(lsn);
+
+        // Can't move backwards.
+        if lsn < old {
+            warn!(
+                "attempted to move last valid LSN backwards (was {}, new {})",
+                old, lsn
+            );
```

Comment:
Why does the comment say " Can't move backwards"?

In my understanding, this situation can legally happen if we restart streaming from last_record_lsn and process some record that wasn't fully processed before.

Reply:
Ah, you're right, I'll tone it down.